src/main.rs

```diff
@@ -436,14 +447,62 @@ fn main() -> anyhow::Result<()> {
 
     touch_started_file_indicator();
 
+    general_runtime.block_on(async {
+        wait_stop_signal().await;
+        // We should ignore send error here. It will happen if all the system
+        // components are already finished and dropped rx portion of the notification
+        // channel, which is totally fine in this case.
+        let _ = shutdown_flag_tx.send(true);
+    });
+
     for handle in handles.into_iter() {
         log::debug!(
             "Waiting for thread {} to finish",
             handle.thread().name().unwrap()
         );
         handle.join().expect("thread is not panicking")?;
     }
-    drop(toc_arc);
-    drop(settings);
+    if let Ok(toc) = wait_unwrap(toc_arc, Duration::from_secs(30)) {
+        drop(toc);
+    }
     Ok(())
 }
+
+/// Blocks current thread for up to specified amount of time to become the single referent
+/// of an [`Arc`] and returns it.
+///
+/// This methods move a value out from an [`Arc`] when there is only one reference left
+/// and it is safe to do so. `Err()` is returned if an [`Arc`] still has more than 1 reference
+/// after given duration is passed.
+fn wait_unwrap<T>(mut input: Arc<T>, duration: Duration) -> Result<T, ()> {
```

Comment:
I think `timeout` is a better name.

```suggestion
fn wait_unwrap<T>(mut input: Arc<T>, timeout: Duration) -> Result<T, ()> {
```

Reply:
done